Hi. Thank you for your contribution!

Tests crash on this PR, for every python version:

```
==== short test summary info 
FAILED tests/test_callbacks.py::test_spawn_n - ZeroDivisionError: division by zero
FAILED tests/test_callbacks.py::test_map_n - TypeError: object list can't be used in 'await' expression
FAILED tests/test_callbacks.py::test_itermap[None] - TypeError: object list can't be used in 'await' expression
FAILED tests/test_callbacks.py::test_itermap[0.1] - TypeError: object list can't be used in 'await' expression
FAILED tests/test_callbacks.py::test_itermap[2] - TypeError: object list can't be used in 'await' expression
FAILED tests/test_map.py::test_itermap - TypeError: object list can't be used in 'await' expression
FAILED tests/test_spawn.py::test_spawns_behaviour - assert 5 == 0
==== 7 failed, 11 passed in 18.51s 
```

--- a/asyncio_pool/base_pool.py
+++ b/asyncio_pool/base_pool.py
@@ -106,7 +106,7 @@ class BaseAioPool(object):
         res, exc, tb = None, None, None
         try:
             res = await coro
-        except Exception as _exc:
+        except BaseException as _exc:
             exc = _exc
             tb = traceback.format_exc()
         finally:
@@ -140,7 +140,7 @@ class BaseAioPool(object):
         acq_error = False
         try:
             await self.semaphore.acquire()
-        except Exception as e:
+        except BaseException as e:
             acq_error = True
             if not future.done():
                 future.set_exception(e)

--- a/asyncio_pool/base_pool.py
+++ b/asyncio_pool/base_pool.py
@@ -182,7 +182,7 @@ class BaseAioPool(object):
         self._waiting[future] = self.loop.create_future()  # as a placeholder
         return await self._spawn(future, coro, cb=cb, ctx=ctx)
 
-    async def spawn_n(self, coro, cb=None, ctx=None):
+    def spawn_n(self, coro, cb=None, ctx=None):
         '''Creates waiting task for given `coro` regardless of pool space. If
         pool is not full, this task will be executed very soon. Main difference
         is that `spawn_n` does not block and returns future very quickly.
@@ -203,7 +203,7 @@ class BaseAioPool(object):
         '''
         return await (await self.spawn(coro, cb, ctx))
 
-    async def map_n(self, fn, iterable, cb=None, ctx=None):
+    def map_n(self, fn, iterable, cb=None, ctx=None):
         '''Creates coroutine with `fn` function for each item in `iterable`,
         spawns each of them with `spawn_n`, returning futures.
 
@@ -211,7 +211,7 @@ class BaseAioPool(object):
         '''
         futures = []
         for it in iterable:
-            fut = await self.spawn_n(fn(it), cb, ctx)
+            fut = self.spawn_n(fn(it), cb, ctx)
             futures.append(fut)
         return futures
 

--- a/asyncio_pool/results.py
+++ b/asyncio_pool/results.py
@@ -14,7 +14,7 @@ def result_noraise(future, flat=True):
     try:
         res = future.result()
         return res if flat else (res, None)
-    except Exception as exc:
+    except BaseException as exc:
         return exc if flat else (None, exc)
 
 

--- a/examples/_usage.py
+++ b/examples/_usage.py
@@ -110,7 +110,7 @@ async def callbacks_usage():
         #   results.append(getres.flat(fut))
         try:
             results.append(fut.result())
-        except Exception as e:
+        except BaseException as e:
             results.append(e)
 
     # First error happens for n == 0 in wrk, exception of it is passed to

--- a/tests/test_base.py
+++ b/tests/test_base.py
@@ -86,18 +86,27 @@ async def test_cancel():
         await aio.sleep(0.5)
         return 1
 
-    pool = AioPool(size=2)
+    async def wrk_safe(*arg, **kw):
+        try:
+            await aio.sleep(0.5)
+        except aio.CancelledError:
+            await aio.sleep(0.1)  # simulate cleanup
+            pass
+        return 1
+
+    pool = AioPool(size=5)
 
     f_quick = pool.spawn_n(aio.sleep(0.15))
-    f12 = await pool.spawn(wrk()), pool.spawn_n(wrk())
-    f35 = pool.map_n(wrk, range(3))
+    f_safe = await pool.spawn(wrk_safe())
+    f3 = await pool.spawn(wrk())
+    pool.spawn_n(wrk())
+    f567 = pool.map_n(wrk, range(3))
 
     # cancel some
     await aio.sleep(0.1)
-    cancelled, results = await pool.cancel(f12[0], f35[2])  # running and waiting
-    assert 2 == cancelled  # none of them had time to finish
-    assert 2 == len(results) and \
-        all(isinstance(res, aio.CancelledError) for res in results)
+    cancelled, results = await pool.cancel(f3, f567[2])  # running and waiting
+    assert cancelled == len(results) == 2  # none of them had time to finish
+    assert all(isinstance(res, aio.CancelledError) for res in results)
 
     # cancel all others
     await aio.sleep(0.1)
@@ -106,11 +115,12 @@ async def test_cancel():
     assert f_quick.done() and f_quick.result() is None
 
     cancelled, results = await pool.cancel()  # all
-    assert 3 == cancelled
-    assert len(results) == 3 and \
-        all(isinstance(res, aio.CancelledError) for res in results)
+    assert cancelled == len(results) == 4
+    assert f_safe.done() and f_safe.result() == 1  # could recover
+    # the others could not
+    assert sum(isinstance(res, aio.CancelledError) for res in results) == 3
 
-    assert await pool.join()  # joins successfully
+    assert await pool.join()  # joins successfully (basically no-op)
 
 
 @pytest.mark.asyncio

--- a/tests/test_base.py
+++ b/tests/test_base.py
@@ -26,7 +26,7 @@ async def test_concurrency():
 
     pool_size = 5
     async with AioPool(size=pool_size) as pool:
-        futures = await pool.map_n(wrk, todo)
+        futures = pool.map_n(wrk, todo)
 
         await aio.sleep(0.01)
 
@@ -70,7 +70,7 @@ async def test_outer_join():
     loop = aio.get_event_loop()
     pool = AioPool(size=100)
 
-    tasks = await pool.map_n(inner, todo)
+    tasks = pool.map_n(inner, todo)
     joined = [loop.create_task(outer(j, pool)) for j in to_release]
     await pool.join()
 
@@ -88,9 +88,9 @@ async def test_cancel():
 
     pool = AioPool(size=2)
 
-    f_quick = await pool.spawn_n(aio.sleep(0.15))
-    f12 = await pool.spawn(wrk()), await pool.spawn_n(wrk())
-    f35 = await pool.map_n(wrk, range(3))
+    f_quick = pool.spawn_n(aio.sleep(0.15))
+    f12 = await pool.spawn(wrk()), pool.spawn_n(wrk())
+    f35 = pool.map_n(wrk, range(3))
 
     # cancel some
     await aio.sleep(0.1)

